docs/source/manual/validation.rst

```diff
@@ -0,0 +1,278 @@
+
+Going back to our favorite ``Person`` class. Let's say we initially coded it such that ``name`` has
+to be non-empty, but realized that business requirements needs the max length to be no more than 5.
+Instead of blowing away our current version of our API and creating angry clients, we can accept
+both versions of the API but at different endpoints.
+
+.. code-block:: java
+
+    public interface Version1Checks { }
+
+    public interface Version2Checks { }
+
+    public class Person {
+        @NotEmpty(groups = Version1Checks.class)
+        @Length(max = 5, groups = Version2Checks.class)
+        private String name;
+
+        @JsonCreator
+        public Person(@JsonProperty("name") String name) {
+            this.name = name;
+        }
+
+        @JsonProperty
+        public String getName() {
+            return name;
+        }
+    }
+
+    @Path("/person")
+    @Produces(MediaType.APPLICATION_JSON)
+    public class PersonResource {
+        @POST
+        @Path("/v1")
+        public void createPersonV1(@Validated(Version1Checks.class) Person person) {
+        }
+
+        @POST
+        @Path("/v2")
+        public void createPersonV2(@Validated({Version1Checks.class, Version2Checks.class}) Person person) {
+        }
+    }
+
+Now, when clients hit ``/person/v1`` the ``Person`` entity will be checked by all the constraints
+that are a part of the ``Version1Checks`` group. If ``/person/v2`` is hit, then all the validations
+are performed.
+
+.. note::
+
+    Since interfaces can inherit other interfaces, ``Version2Checks`` can extend ``Version1Checks``
+    and wherever ``@Validated(Version2Checks.class)`` is used, version 1 constraints are checked
+    too.
+
+.. _man-validation-testing:
+
+Testing
+=======
+
+It is critical to test the constraints so that you can ensure the assumptions about the data hold
+and see what kinds of error messages clients will receive for bad input. The recommended way for
+testing annotations is through :ref:`Testing Resources <man-testing-resources>`, as Dropwizard does
+a bit of magic behind the scenes when a constraint violation occurs to set the response's status
+code and ensure that the error messages are user friendly.
+
+.. code-block:: java
+
+    @Test
+    public void personNeedsAName() {
+        // Tests what happens when a person with a null name is sent to
+        // the endpoint.
+        final Response post = resources.client()
+                .target("/person/v1").request()
+                .post(Entity.json(new Person(null)));
+
+        // Clients will recieve a 422 on bad request entity
+        assertThat(post.getStatus()).isEqualTo(422);
+
+        // Check to make sure that errors are correct and human readable
+        ValidationErrorMessage msg = post.readEntity(ValidationErrorMessage.class);
+        assertThat(msg.getErrors())
+                .containsOnly("name may not be empty");
+    }
+
+ .. _man-validation-extending:
```

Comment:
A subtle, but important omission. There is a whitespace at the start of the line and `.. _man-validation-extending:` is treated as a part the source code block.